examples/sha256/table16.rs

```diff
@@ -158,93 +161,158 @@ impl<F: FieldExt> Table16Chip<F> {
+    }
+
+    /// Exits out of the existing namespace.
+    ///
+    /// Not intended for downstream consumption; use [`Layouter::namespace`] instead.
+    fn pop_namespace(&mut self, _gadget_name: Option<String>) {
+        // TODO
     }
 }
 
-impl<F: FieldExt> Sha256Instructions for Table16Chip<F> {
+impl<F: FieldExt, L: Layouter<Field = F>> Sha256Instructions for Table16Config<'_, F, L> {
     type State = State;
     type BlockWord = BlockWord;
 
     fn zero() -> Self::BlockWord {
         BlockWord::new(0)
     }
 
-    fn initialization_vector(layouter: &mut impl Layouter<Self>) -> Result<State, Error> {
-        let config = layouter.config().clone();
-        config.compression.initialize_with_iv(layouter, IV)
+    fn initialization_vector(&mut self) -> Result<State, Error> {
+        let configured = self.configured().compression.clone();
+
+        let mut compression = CompressionConfig::<'_, F, L> {
+            configured,
+            layouter: self.layouter(),
+            marker: PhantomData,
+        };
+        compression.initialize_with_iv(IV)
     }
 
-    fn initialization(
-        layouter: &mut impl Layouter<Table16Chip<F>>,
-        init_state: &Self::State,
-    ) -> Result<Self::State, Error> {
-        let config = layouter.config().clone();
-        config
-            .compression
-            .initialize_with_state(layouter, init_state.clone())
+    fn initialization(&mut self, init_state: &Self::State) -> Result<Self::State, Error> {
+        let configured = self.configured().compression.clone();
+
+        let mut compression = CompressionConfig::<'_, F, L> {
+            configured,
+            layouter: self.layouter(),
+            marker: PhantomData,
+        };
+        compression.initialize_with_state(init_state.clone())
     }
 
     // Given an initialized state and an input message block, compress the
     // message block and return the final state.
     fn compress(
-        layouter: &mut impl Layouter<Self>,
+        &mut self,
         initialized_state: &Self::State,
         input: [Self::BlockWord; super::BLOCK_SIZE],
     ) -> Result<Self::State, Error> {
-        let config = layouter.config().clone();
-        let (_, w_halves) = config.message_schedule.process(layouter, input)?;
-
-        config
-            .compression
-            .compress(layouter, initialized_state.clone(), w_halves)
+        let configured = self.configured().clone();
+
+        let mut message_schedule = MessageScheduleConfig::<'_, F, L> {
+            configured: configured.message_schedule.clone(),
+            layouter: self.layouter(),
+            marker: PhantomData,
+        };
+
+        let (_, w_halves) = message_schedule.process(input)?;
+
+        let mut compression = CompressionConfig::<'_, F, L> {
+            configured: configured.compression,
+            layouter: self.layouter(),
+            marker: PhantomData,
+        };
+        compression.compress(initialized_state.clone(), w_halves)
```

Comment:
Similarly this would be:
```suggestion
        let (_, w_halves) = self.message_schedule.process(input)?;
        self.compression.compress(initialized_state.clone(), w_halves)
```